### PR TITLE
feat: add a (readonly) local document store

### PR DESCRIPTION
**Pull request by bjoerge**
Note: this supersedes #25 and includes many of the changes in it.

This introduces a new read only local document store, which is a lot simpler in nature than the optimistic store, and very useful for cases where you don't need to be able to optimistically apply edits, but still want to apply edits to- and receive updates while maintaining a local copy of the latest version of one or more documents. With this store you can listen for updates to snapshots of documents (by id) and get listener connection reuse/deduping/multiplexing listener events and shareing instances of local documents between all interested parties in the application.

Included with this PR are also sequentializing of listener events and error recovery logic in case of lost events added in #25.

Here's how to set it up and use it:
```ts
import {createClient} from '@sanity/client'
import {
  createDocumentEventListener,
  createDocumentLoader,
  createDocumentUpdateListener,
  createReadOnlyStore,
  createSharedListener,
} from '@sanity/mutate/_unstable_store'

const client = createClient({
  /* client config */
})

// Create a document loader (data loader that will batch requests for individual documents)
const loadDocument = createDocumentLoader({client})

const sharedListenerEvents = createSharedListener({
  client,
  // optional filter. Must match the doucment types you want to subscribe to.
  // defaults to `*`
  // filter: `_type == 'some-type'`
})

const listenDocumentEvents = createDocumentEventListener({
  // pass a function that will load a document by id
  loadDocument,
  // this must be a stream of listener events, in our case we're using a shared listener for the document type(s) we're interested in
  listenerEvents: sharedListenerEvents,
})

const listenDocumentUpdates = createDocumentUpdateListener({
  // pass it our shared document event listener
  listenDocumentEvents,
})

// create the store
const store = createReadOnlyStore(listenDocumentUpdates)

// we can now subscribe to documents
store.listenDocument('foo').subscribe(update => {
  console.log(update.event, update.snapshot)
})

// we can now subscribe to documents
store.listenDocuments(['foo', 'bar']).subscribe(([fooUpdate, barUpdate]) => {
  // this will be called with latest foo and bar when either change
  console.log(fooUpdate.snapshot, barUpdate.snapshot)
})

// We can also use the document loader and the listeners created above independently if we want to:
listenDocumentEvents('foo').subscribe(event => {
  console.log(event.type) // logs 'reconnect', 'sync' or 'mutation'
})

// We can also subscribe to every listener event matching the filter
// this will establish an SSE connection upon first publish, share the underlying event stream between all subscribers
// and disconnect the SSE connection when the last subscriber unsubscribes
sharedListenerEvents.subscribe(event => {
  console.log(event.type) // logs 'reconnect', 'welcome' or 'mutation'
})

// Load a document by id from the /doc endpoint. Not cached, only batched and deduped dataloader style
loadDocument('bar').subscribe(result => {
  if (result.accessible) {
    console.log(result.document)
  } else {
    console.log('Document %s not accessible: %s', result.id, result.reason)
  }
})

```

Notable changes:
- Renames the existing `createContentLakeStore`  => `createOptimisticStore`
- Refactors creator functions for document loader and shared listeners so they can more easily be re-used.
- Also adds a `listenIdSet(<groqFilter>)` lightweight listener function for subscribing to the (unordered) set of document ids that matches a certain groq-filter.

### CODE REVIEW

**Review by stipsan — COMMENTED** [review]
Stellar work here! 💖  It's a great refactor, cleanup, hardening of critical APIs, streamlining of the API surface and the new functionality will be a joy to adopt in our state machines 🥳 
The only reason I'm hesitant to approve is the dependency on `@sanity/client` moving to a prerelease dist-tag, can we land the changes you need in `@sanity/client@latest` before we merge and release this PR? 🤔

File: package.json [hunk 1]
```diff
@@ -89,7 +89,7 @@
   },
   "browserslist": "extends @sanity/browserslist-config",
   "dependencies": {
-    "@sanity/client": "^6.22.3",
+    "@sanity/client": "6.22.5-beta.1",
```
[the remark this answers is not in the document]
Comment (bjoerge) [reply package.json:92]: Ah, dang – I meant to remove this before opening this PR – I was just testing using eventsource@3. I'll revert this.

File: src/apply/patch/__test__/applyPatch.bench.ts
[the remark this answers is not in the document]
Comment (bjoerge) [reply src/apply/patch/__test__/applyPatch.bench.ts:1]: Removed, thanks!

File: package.json (hunk 1 above)
[the remark this answers is not in the document]
Comment (bjoerge) [reply package.json]: Reverted.

**Review by stipsan — APPROVED** [review]

**Pull request merged**